### ISSUE TITLE
Perhaps some refactoring around `FilesGenerator` / `CodeInjector`?

### ISSUE THREAD

**Issue by vsajip** [javacc21/javacc21#78]
Every one of the `CodeInjector`'s constructor parameters comes from the `Grammar` instance:
https://github.com/javacc21/javacc21/blob/91360976ac5507ce08270e3c1d356f4be99256b7/src/java/com/javacc/output/java/FilesGenerator.java#L72-L78

Given that we'll need to manage injections independently of target language, it seems that `CodeInjector` could do with moving out of the `output.java` subpackage and perhaps into the same package as `Grammar`, and over time making it more language-agnostic.  As we discussed earlier, `FilesGenerator` could do with being more data-driven and less copy/paste/modify, and we might also want to move this to be in the same package as `Grammar`. That leaves the `output.java` subpackage would only contain `JavaCodeUtils` and `JavaFormatter`, which seems appropriate as they are Java-specific. Obviously `CodeInjector` is currently Java-specific, but that could change over time and perhaps some of its work could be delegated to language-specific helper code. What do you think about this idea?

**Comment by revusky** [comment]
> What do you think about this idea?

To be honest, I don't really know. Sometimes I write things in a very bloody minded way, just taking something and copy-paste-modifying (which I know is evil in principle....) but sometimes I find that just doing that and then just stepping back, sometimes you see your way clear to consolidating the repeated elements. It might well be that the best thing to do at this precise point is just to create a com.javacc.output.python package and stick in the FilesGenerator and CodeInjector classes that are largely duplicates of the ones in com.javacc.output.java. I mean to say, just to do things in a totally naive bloody-minded way, and then later revisit how to consolidate the duplicated logic. Well, I assume that renaming FilesGenerator to JavaFilesGenerator and CodeInjector to JavaCodeInjector would make sense. And then in the equivalent python package, we'd have PythonFilesGenerator and PythonCodeInjector.

I guess that they'd eventually have a common superclasses, like AbstractFilesGenerator and AbstractCodeInjector maybe, which would consolidate duplicated logic. But I think it might be hard to figure it out in advance. I think for me it would be. So I think what I would most likely do is just duplicate everything with Python and then look for the opportunities to consolidate the duplicated logic later.

Well, the above is just my wooly-minded thoughts. Obviously, if you can get there via a different route....  Finally the proof of the pudding is in the eating, after all... Or to put it another way, we can both visualize the desired endpoint, more or less anyway, so what intermediate steps you take to get there.... well... as long as we get there...

**Comment by vsajip** [comment]
> It might well be that the best thing to do at this precise point is just to create a com.javacc.output.python package and stick in the FilesGenerator and CodeInjector classes that are largely duplicates of the ones in com.javacc.output.java

I'd prefer to avoid that if possible, and I don't mind having a go at the refactoring.

The thing is that `CodeInjector` consists of two parts: collecting the injections and then writing the node classes that contain them. The first part is generic, the second part needs to be language-specific.

I can leave the `FilesGenerator` as it is for now, with perhaps two hardcoded paths for Java and Python output, and come back to it later.

I'd add public methods to `CodeInjector` to make e.g. the extends lists, and so on, available to e.g. template logic. For example, to generate node classes in Python, I would expect to use some template logic and/or helper methods to generate e.g. the superclasses for AST nodes correctly. This information is currently held internally in the injector instance.

Obviously I'm not suggesting making those changes in this repo - I will play with it privately until I have a reasonable working solution, which is the point at which I create a branch, publish the changes and create a PR for your review.

**Comment by vsajip** [comment]
Of course, I would have to make `CodeInjector`, its constructor and some methods `public`.

**Comment by revusky** [comment]
> Of course, I would have to make `CodeInjector`, its constructor and some methods `public`.

Well, really, I don't care very much about that. I tend to give things only the visibility level they need to have. But if something has to be public, then.... Generally, if you're willing to attack the problem, just go for it. All I was saying, I guess, was that sometimes it takes two passes (or more) to get things really right so a crude initial copy-paste-modify version is not necessarily so terrible. Sometimes there is a danger of getting into paralysis by (over) analysis. But, I guess, ideally, when we have gone from one target language to two, it should then be pretty obvious for anybody eyeballing the code how go from two to three...

**Comment by revusky** [comment]
I'm closing this issue mostly for housekeeping since all new ongoing development is at https://github.com/congo-cc/congo-parser-generator THe truth of the matter is that I think I'm going to rewrite all the CodeInjector stuff. Actually, that code is not very changed from 2008 when I forked off FreeCC!
